test/indexOf.js

```diff
@@ -3,25 +3,13 @@
 var S = require('..');
 
 var eq = require('./internal/eq');
-var throws = require('./internal/throws');
 
 
 test('indexOf', function() {
 
   eq(typeof S.indexOf, 'function');
   eq(S.indexOf.length, 2);
-
-  throws(function() { S.indexOf('x', null); },
-         TypeError,
-         'Invalid value\n' +
-         '\n' +
-         'indexOf :: a -> List a -> Maybe Integer\n' +
-         '                ^^^^^^\n' +
-         '                  1\n' +
-         '\n' +
-         '1)  null :: Null\n' +
-         '\n' +
-         'The value at position 1 is not a member of ‘List a’.\n');
+  eq(S.indexOf.toString(), 'indexOf :: a -> List a -> Maybe Integer');
```

Comment:
The List type attempts to unify [`$.Array`][1] and [`$.String`][2]. It doesn't have anything to do with linked lists. It's a confusing type, certainly!

> ### List
>
> The List type constructor enables type signatures to describe ad hoc polymorphic functions which operate on either [`Array`][1] or [`String`][2] values.
>
> Mental gymnastics are required to treat arrays and strings similarly. `[1, 2, 3]` is a list containing `1`, `2`, and `3`. `'abc'` is a list containing `'a'`, `'b'`, and `'c'`. But what is the type of `'a'`? `String`, since JavaScript has no Char type! Thus:
>
>     'abc' :: String, List String, List (List String), ...
>
> Every member of `String` is also a member of `List String`! This affects the interpretation of type signatures. Consider the type of `indexOf`:
>
>     a -> List a -> Maybe Integer
>
> Assume the second argument is `'hello' :: List String`. `a` must then be replaced with `String`:
>
>     String -> List String -> Maybe Integer
>
> Since `List String` and `String` are interchangeable, the former can be replaced with the latter:
>
>     String -> String -> Maybe Integer
>
> It's then apparent that the first argument needn't be a single-character string; the correspondence between arrays and strings does not hold.

As for `S.indexOf` and `S.lastIndexOf`, I'd really like to remove these functions. Addressing array elements by index is not a practice we should encourage.


[1]: https://github.com/sanctuary-js/sanctuary-def#Array
[2]: https://github.com/sanctuary-js/sanctuary-def#String